### PR TITLE
Alert View Model

### PR DESCRIPTION
**Pull request by coaguila**
# Alert View Model

## Description
This PR introduces a view model for the Alerts feature that will be implemented. It closes issue #118 

## Changes
Added a view model to handle the alert feature that will allow a screen to create and delete their alerts as well as get all the alerts including those from other users.

#### Added
- `app/src/main/java/com/android/periodpals/model/alert/AlertViewModel.kt`
- `app/src/test/java/com/android/periodpals/model/alert/AlertViewModelTest.kt`

#### Modified
- `app/src/main/java/com/android/periodpals/model/alert/Alert.kt`
- `app/src/main/java/com/android/periodpals/model/alert/AlertDto.kt`
- `app/src/main/java/com/android/periodpals/model/alert/AlertModel.kt`
- `app/src/main/java/com/android/periodpals/model/alert/AlertModelSupabase.kt`
- `app/src/test/java/com/android/periodpals/model/alert/AlertModelSupabaseTest.kt`

### CODE REVIEW

File: app/src/test/java/com/android/periodpals/model/alert/AlertModelSupabaseTest.kt
Comment (francelu) [inline app/src/test/java/com/android/periodpals/model/alert/AlertModelSupabaseTest.kt:1]: - It has the wrong package (to say none at the top of the file).
- Nitpick: in the `companion object`, you could clean up the vals with `const val ID` (capital val name), ...

File: app/src/main/java/com/android/periodpals/model/alert/AlertViewModel.kt [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+      alertModelSupabase.getAlert(
+          idAlert = idAlert,
+          onSuccess = { fetched ->
+            Log.d(TAG, "getAlert: Success")
+            alert = fetched
+          },
+          onFailure = { e -> Log.e(TAG, "getAlert: fail to get alert: ${e.message}") })
+    }
+    return alert
+  }
```
Comment (francelu) [inline app/src/main/java/com/android/periodpals/model/alert/AlertViewModel.kt:59]: This isn't tested, please do so.

File: app/src/main/java/com/android/periodpals/model/alert/AlertViewModel.kt [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+          // UserViewModel.currentUser?.uid
+          cond = { eq("uid", uid) },
+          onSuccess = { alerts ->
+            Log.d(TAG, "getMyAlerts: Success")
+            alertsList = alerts
+          },
+          onFailure = { e -> Log.e(TAG, "getMyAlerts: fail to get alerts: ${e.message}") })
+    }
+    return alertsList
+  }
```
Comment (francelu) [inline app/src/main/java/com/android/periodpals/model/alert/AlertViewModel.kt:100]: Same. This isn't tested, please do so.

File: app/src/main/java/com/android/periodpals/model/alert/AlertViewModel.kt [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+    viewModelScope.launch {
+      alertModelSupabase.updateAlert(
+          alert = alert,
+          onSuccess = {
+            Log.d(TAG, "updateAlert: Success")
+            getAllAlerts()
+          },
+          onFailure = { e -> Log.e(TAG, "updateAlert: fail to update alert: ${e.message}") })
+    }
+  }
```
Comment (francelu) [inline app/src/main/java/com/android/periodpals/model/alert/AlertViewModel.kt:117]: Same. This isn't tested, please do so.

File: app/src/test/java/com/android/periodpals/model/alert/AlertViewModelTest.kt
Comment (francelu) [inline app/src/test/java/com/android/periodpals/model/alert/AlertViewModelTest.kt:1]: - Good implementation.
- However, missing tests for `getAlert()`, `getAlertsByUser()` and `updateAlert()`.

**Review by francelu — CHANGES_REQUESTED** [review]
## Summary
Good implementation. However, as mentioned in the comments, we still need more tests for the alert VM. It would be good to have a function to fetch pals alerts list too.

## Important
#### Code Quality
- Good implementation. Code is clear and consistent. Good job.
- Missing a function to fetch pals alerts list.

#### Functionality
- Overall good.
- As mentioned, missing function for the pals alerts list.
- Missing some edge cases tests for alert VM.

#### Performance
- Unchanged.

#### Security
- Unchanged.

## Testing
- Missing tests for a couple of functions of alert VM. See comments for more details.

## Documentation
- Adapted and added documentation as needed. Well done !

## Suggestions
- See comments.

## Steps before PR approved
- Add a function to fetch pals alerts list.
- Add edge cases tests.
- Read comments and replay / resolve as necessary.

**Review by taghizadlaura — CHANGES_REQUESTED** [review]
## Summary  
The `AlertViewModel` class provides a structured approach to managing alert data, including methods for creating, retrieving, updating, and deleting alerts. The associated test class validates the behavior of these methods by mocking the `AlertModelSupabase` repository and testing the key functionalities such as successful and failed alert operations.

## Important  

#### Code Quality  
LGTM :
- **AlertViewModel**:  
  The `AlertViewModel` uses `mutableStateOf` for alerts and exposes it via a public `State`. Its methods (e.g., `createAlert`, `updateAlert`, `deleteAlert`) interact with the `AlertModelSupabase` repository to handle background operations.  

- **AlertViewModelTest**:  
  The test class effectively mocks the repository, testing core functionalities such as creating, updating, deleting, and retrieving alerts. It covers success and failure scenarios, ensuring proper state updates.

#### Functionality  

- **AlertViewModel**:  
  The methods integrate with the `AlertModelSupabase` repository for alert management, ensuring data is correctly loaded, updated, and deleted. Further testing is needed to ensure methods like `getAlert`, `getAllAlerts`, and `getAlertsByUser` return up-to-date data after operations.

- **AlertViewModelTest**:  
  The tests validate key ViewModel methods, simulating repository behavior. Additional edge case tests (e.g., invalid data, network failures) would enhance coverage.


#### Performance
LGTM :
There are no significant performance concerns introduced by the `AlertViewModel`. 

#### Security 
LGTM :
No security vulnerabilities were introduced..

## Testing  
The `AlertViewModelTest` effectively tests the major operations provided by the `AlertViewModel`. The test class mocks the repository to simulate various scenarios:
- **Create**: Verifies successful creation and failure due to network issues.
- **Update**: Tests the update functionality.
- **Delete**: Confirms that an alert is deleted successfully.
- **Retrieve**: Verifies the retrieval of alerts and ensures that the ViewModel updates correctly.

While the tests effectively cover the core functionalities, I feel that the  the `getAlert`, `getAlertsByUser` and `getAllAlerts` functions in the AlertViewModel do have missing tests. 

## Documentation  
LGTM :
Both the ViewModel and the test class are well-documented. However, adding further comments to explain certain design decisions (such as why some methods return values synchronously) would enhance the maintainability of the code.

## Suggestions  
- Enhance test coverage by adding tests for edge cases like invalid data, network timeouts, malformed responses, and handling missing alert IDs.
- Ensure the `AlertViewModel` properly handles network failures and other edge cases during interactions with the repository.


## Steps Before PR Approval  
- Please add tests for `getAlert`, `getAllAlerts`, and `getAlertsByUser` to ensure they return up-to-date data after operations.  
- If you can, please include edge case tests (e.g., invalid data, network issues) to improve ViewModel handling.

**Comment by francelu** [comment]
Hey, have you checked the supabase table for alerts ?  
I think it needs a primary key at least.  
image

**Review by taghizadlaura — APPROVED** [review]
## Summary
LGTM :
The `AlertViewModelTest` class effectively tests the alert operations in the `AlertViewModel`, including creation, deletion, retrieval, and updates. Mocks for the `AlertModelSupabase` repository validate the core functionalities.

## Important

#### Code Quality
LGTM :
- **AlertViewModelTest**:  
  The tests have been expanded to cover a wider range of scenarios, including successful and failed alert creation, update, deletion, and retrieval. The tests for methods like `getAlert`, `getAllAlerts` and `getAlertsByUser` ensure the ViewModel handles data properly under both success and failure conditions.

#### Functionality
LGTM :
- **AlertViewModelTest**:  
  Tests now cover additional scenarios such as handling failures for alert retrieval and deletion. The `deleteAlertNullIdFailure` test ensures the ViewModel handles alerts with null IDs correctly.  
  The new tests for `getAlert`,  `getAllAlerts` and `getAlertsByUser` validate their proper functionality with success and failure cases.

#### Performance
No performance concerns have been introduced.

#### Security
No security issues were introduced.

## Testing
Key operations are tested with mock failures and success cases.

## Documentation
The tests are clear, but further documentation of specific edge case handling could enhance maintainability.

## Suggestions
Consider enhancing the tests to cover additional edge cases, such as network issues or malformed responses, to ensure more robust error handling in the ViewModel.

## Steps before PR approved
LGTM

File: app/src/main/java/com/android/periodpals/model/alert/AlertViewModel.kt [hunk 4]
```diff
@@ -77,6 +77,11 @@ class AlertViewModel(private val alertModelSupabase: AlertModelSupabase) : ViewM
     return alertsList
   }
 
+  fun getPalAlerts(uid: String): List<Alert>? {
+    val palAlertList: List<Alert>? = getAllAlerts()
+    return palAlertList?.filter { it.uid != uid }
+  }
+
```
Comment (francelu) [inline app/src/main/java/com/android/periodpals/model/alert/AlertViewModel.kt:84]: Missing documentation :)

**Review by francelu — APPROVED** [review]
## Summary
LGTM ! You did the necessary changes. Minor nitpicks.

## Important
#### Code Quality
- Code is clean, readable and consistent. Code is simple and not overwhelming :)

#### Functionality
- All good. You implemented well the requested changes notably in the tests !

#### Performance
- Unchanged.

#### Security
- Unchanged.

## Testing
- Improved ! As mentioned already, you completed the tests as requested.

## Documentation
- All good, just missing one on the new `getPalsAlerts()` (check code comment).

## Suggestions
- Add documentation for `getPalAlerts()`.
- **Nitpick**: `AlertModeSupabaseTest` class still has no package at the beginning of the file :')

## Steps before PR **merged**
See the above mentioned suggestions, refer to code comments, else all good. Thanks for the hard work ! You've been working a lot on the VMs.


- Add documentation as mentioned in code comment.
